packages/compat/src/transaction.ts

```diff
@@ -87,20 +85,10 @@ function convertSignatures(
     }, {});
 }
 
-export function fromVersionedTransactionWithBlockhash(
-    transaction: SignedVersionedTransaction,
-    lastValidBlockHeight: bigint
-): Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime & ITransactionWithSignatures;
-
-export function fromVersionedTransactionWithBlockhash(
-    transaction: VersionedTransaction,
-    lastValidBlockHeight: bigint
-): Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime;
-
 export function fromVersionedTransactionWithBlockhash(
     transaction: VersionedTransaction,
     lastValidBlockHeight: bigint
-): Transaction {
+): Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime {
```

Comment:
So with this change it will copy over my signatures but I won't get a `ITransactionWithSignatures` type response, right?
You might be able to wrangle it with generics, ie:

```typescript
type OldTransactionUnsigned = VersionedTransaction;
type OldTransactionSigned = VersionedTransaction & { signatures: [] };

type NewTransactionUnsigned = Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime
type NewTransactionSigned = Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime & ITransactionWithSignatures

// Psuedo-code
fn<TOldTransaction extends OldTransactionUnsigned | OldTransactionSigned>(old: TOldTransaction): TOldTransaction extends OldTransactionUnsigned ? NewTransactionUnsigned : NewTransactionSigned
```

Reply:
I tried stuff like this and the problem is that it can't inspect the old transaction and determine whether it has signatures. I guess because of the lack of a Typescript runtime. Which is why I ended up introducing `SignedVersionedTransaction`

Reply:
Ohhhhhh right I _did_ totally read this when you first wrote it a few commits ago, but now I get it

Reply:
This actually works though (or appears to)!
[playground link](https://www.typescriptlang.org/play?#code/PTAEBkFMHMEMGMCeAoALogDpUA1SAnAZwEsB7AO0gBMAVfWcwhVM80AXlAG9lQ-RijVPgCu8FhUIAuUIWGDoAbQC6AGl78AZpEgAFWIgIy5+Bev6gARgBtS8ANYALWIUfH55aOf4lo5WKgi+JCEAPwy5CIAtpYEKsgAvsjIIKAAogAeWKZRkOSosNZomNh0DEzirBzcGnyCJmISjO6mnvFJ6FigAJJljMysAOrEqI4AYjr6hvjVPBbaegZGsh5eicVdvfT9lRTDowBCtg7OruDE2iy5s7VWx04ubiutax0lPX0VTfuOAMrEfgCQRCNwsvn8gWC0meChU6iSG2wAHlrLRtl9WABVRgAyhUap4IisaifAYUADciNAKLR5TJ5H+fmoBIIJAoJPR9NAADJuLJcUCoRForF8CpQAlKVSAHKQADupN25Gx4OZnEVTR5H05Sp+E0W0y1WzpupGjiOdgeZwukCukBl8o1rEZeOqTooRvd5D1kyWM15xp23zNFpOj3Ol2I1wDXp+LsFIWSmhE5CVoE05AAPDQaV7QJAMqg8lRCNTUV6VbjmQAfMu0oPOqtUAB8AApSKiZDnyzqmgBKLu53tVAtF8glutel3M0KgWUK4cUaf4mTzis4pn4ub8YKQtiRazWUAuUAp+zkUhytgn1vd+sYj2j4ulocmprL0CzteLhlN0Crx0f0rTc+3WZB4EkVBQAvaciGqDNW23OohFEJVoUUAByYgMgwtRbgWKZlgwgArUhIAw7w+BsS1TieDDqIcCjElA8DINAOUzVg0tOAQpCBBQxpWHQrCcLw+ZfWmGQSLIpiLAYq06Pk2SfAFSEQhkRQAFZlGYoA)

Reply:
Nice! Seems to work with empty vs non-empty array too which is cool: [playground](https://www.typescriptlang.org/play?#code/PTAEBkFMHMEMGMCeAoALogDpUA1SAnAZwEsB7AO0gBMAVfWcwhVM80AXlAG9lQ-RijVPgCu8FhUIAuUIWGDoAbQC6AGl78AZpEgAFWIgIy5+Bev6gARgBtS8ANYALWIUfH55aOf4lo5WKgi+JDSoOQiALaWBCrIAL7IyCCgAKIAHlimEZDkqLDWaJjYdAxM4qwc3Bp8giZiEozupp6xCehYoACSJYzMrADqxKiOAGI6+ob4lTwW2noGRrIeXvGFHd30veUUg8MAQrYOzq7gxNos2dPVVodOLm5LzSttRV09ZQ27jgDKxH4BQRCVwsvn8gWCoRMChU6gSa2wADkKCkIhh0ABBfD0RAAHhoAD5KooaKpQAA6Ck0FTKRLtbAAeWstE2H1YAFVGH9KFRKngiKxqO8+hQANzw0CM5mlYXkX5+ai8ggkCiClky0AAMm4si5AIhMiR5BRaMQmOxOPCUQIhLiYvFCMgAHchdtyBzQQrOC6Gpq3mrXV8xvNJr6NtKA0NHAc7HcTmdIBdIPand7WHLuZVUxRQ1nyIHxgsplqw1tPpHo0d7qdzsRLsXc1903qQolNCJyK7QJpyHjJbnQJA0qgclRCBKmbn3VyFQAfcdS0tp6dUfEAClITJkND7-oaAEotzvwz7B8PyKP57n0wqAPygB3O3dL+U8g0pp8UKcvqoWYLgtjhNY1igC4oDtvY5CkI6bCgau24Th+bCniOY5HouFDXjyd4Pley6gG+j7HuynIvnuqzIPAkioGEpDXkQlTdquPDJBYAB6N7XLUwj1KwoSKAA5MQaT8Wo1xzBMiz8QAVqQkD8d4fA2DGxwPPxSkOPJ1ygs2fE0nEZEUVRoCOpGdFjpwjHMWAbEcRYXGiK6fGCcJomzAWkwyNJsmaRY6mxqpfk+T4urgiEMiKAArHpe5AA)

I don't think we'll be able to get to differentiating how it actually is on the old transactions though, you need to differentiate all signatures are all-zero arrays vs not. The old signature stuff seems designed to stop us doing this lol

Reply:
ohhh yeah I see what you mean. So if you provide "with signatures" and it's just one single all-zero signature, you'll get back "with signatures" but it will be empty

Reply:
Is it all-zero automatically? In other words, this whole thing isn't ever going to resolve to "unsigned"?

Reply:
I don't know for sure but in my testing all the ways you're meant to add a signature or a signer add an empty signature to the list of signatures. And AFAICT using that index is the only way to actually match an account to a signature, so it's probably a bug if there are any cases where they become disconnected. It's not a type-level guarantee though and you can probably mess it up pretty easily if you move outside of the class methods.

Reply:
Since signatures are fixed length, we could maybe use something like this in our conditional type-handling:

```typescript
type AllZeroSignature = [0, 0, 0, 0];
type SomeSignature = [number, number, number, number];

type ValidSignature = SomeSignature extends AllZeroSignature ? never : SomeSignature;
```

Reply:
but this might not work within another array :(

Reply:
Ah wait I think this is why it wasn't working for me: [playground](https://www.typescriptlang.org/play?#code/PTAEBkFMHMEMGMCeAoALogDpUA1SAnAZwEsB7AO0gBMAVfWcwhVM80AXlAG9lQ-RijVPgCu8FhUIAuUIWGDoAbQC6AGl78AZpEgAFWIgIy5+Bev6gARgBtS8ANYALWIUfH55aOf4lo5WKgi+JCEAPwy5CIAtpYEKsgAvsjIIKAAogAeWKZRkOSosNZomNh0DEzirBzcGnyCJmISjO6mnvFJ6FigAJJljMysAOrEqI4AYjr6hvjVPBbaegZGsh5eicVdvfT9lRTDowBCtg7OruDE2iy5s7VWx04ubiutax0lPX0VTfuOAMrEfgCQRCNwsvn8gWC0meChU6iSG2wAHlrLRtl9WABVRgAyhUap4IisaifAYUADciNAKLR5TJ5H+fmoBIIJAoJPR9NAADJuLJcUCoRForF8CpQAlKVSAHKQADupN25Gx4OZnEVTR5H05Sp+E0W0y1WzpupGjiOdgeZwukCukBl8o1rEZeOqTooRvd5D1kyWM15xp23zNFpOj3Ol2I1wDXp+LsFIWSmhE5CVoE05AAPDQaV7QJAMqg8lRCNTUV6VbjmQAfMu0oPOqtUAB8AApSKiZDnyzqmgBKLu53tVAtF8glutel3M0KgWUK4cUaf4mTzis4pn4ub8YKQtiRazWUAuUAp+zkUhytgn1vd+sYj2j4ulocmprL0CzteLhlN0Crx0f0rTc+3WZB4EkVBQAvaciBkQk2TxPNOG3OohFEJVoUUAByYgMmwtRbgWKZlmwgArUhIGw7w+BsS1TiebC6IcaiwIgoRoNIWDCAAYQoAA3Agi3xTgM1bGCqyIPtKVSfgAD1QmSdi5FAOUzW4+DWWJe8uRQ256mERpWCw3D8MI+ZfWmGRyMo1iLGYq1GIcuyfAFSEQhkRQAFZlDYyDVPUyTeIEoS1XTchWzU0ZuOklIwHk0IgA)

AFAICT as soon as you add the indirection of 'hey typescript this is a VersionedTransaction from somewhere else' and not 'hey typescript here is a VersionedTransaction being explicitly created right here so you can see it' it loses the ability to do anything that involves inspecting the transaction. I don't know if I'm missing something that unlocks this though because even if it doesn't work here because signatures are weird that would be awesome!

Reply:
Ahhh damn, it looks like just casting it to a `VersionedTransaction` from anywhere breaks it. Smh